### ISSUE TITLE
Bug on multi threading

### ISSUE THREAD

**Issue by sthebault** [locie/pySIP#2]
While running the armadillo example "fit bayesian", the chains are running correctly, but at the end the following error appears:

```
Exception ignored in: <function tqdm.__del__ at 0x000002994E768F28>
Traceback (most recent call last):
  File "xx\anaconda3\lib\site-packages\tqdm\_tqdm.py", line 931, in __del__
    self.close()
  File "xx\anaconda3\lib\site-packages\tqdm\_tqdm.py", line 1133, in close
    self._decr_instances(self)
  File "xx\anaconda3\lib\site-packages\tqdm\_tqdm.py", line 496, in _decr_instances
    cls.monitor.exit()
  File "xx\anaconda3\lib\site-packages\tqdm\_monitor.py", line 52, in exit
    self.join()
  File "xx\anaconda3\lib\threading.py", line 1029, in join
    raise RuntimeError("cannot join current thread")
RuntimeError: cannot join current thread
```

**Comment by LoicRaillon** [comment]
Did you install the requirements? Which OS are you using?

**Comment by sthebault** [comment]
Yes I did the python setup.py (that installs the requirements). I'm using Windows x64

**Comment by sthebault** [comment]
I tried to reinstall all the requirements in an empty environment (python 3.7) and I got a different error:

joblib.externals.loky.process_executor._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\site-packages\joblib\externals\loky\process_executor.py", line 418, in _process_worker
    r = call_item()
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\site-packages\joblib\externals\loky\process_executor.py", line 272, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\site-packages\joblib\_parallel_backends.py", line 567, in __call__
    return self.func(*args, **kwargs)
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\site-packages\joblib\parallel.py", line 225, in __call__
    for func, args, kwargs in self.items]
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\site-packages\joblib\parallel.py", line 225, in <listcomp>
    for func, args, kwargs in self.items]
  File "Xx\PycharmProjects\pySIP\pysip\mcmc\hmc.py", line 226, in _sample_chain
    q, s = self._hmc_step(q, stepsize)
  File "Xx\PycharmProjects\pySIP\pysip\mcmc\hmc.py", line 283, in _hmc_step
    state_n, stats = self._transition(state, stepsize)
  File "Xx\PycharmProjects\pySIP\pysip\mcmc\hmc.py", line 327, in _transition
    depth, -1, state_l, stepsize, H0, sum_p_s, sum_w_s, stats
  File "Xx\PycharmProjects\pySIP\pysip\mcmc\hmc.py", line 382, in _build_tree
    state = self._leapfrog_step(state, direction, stepsize)
  File "Xx\PycharmProjects\pySIP\pysip\mcmc\hmc.py", line 470, in _leapfrog_step
    dK = self._hamiltonian.dK(p)
  File "Xx\PycharmProjects\pySIP\pysip\mcmc\hamiltonian.py", line 78, in dK
    return sla.cho_solve((self._cholM, True), p)
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\site-packages\scipy\linalg\decomp_cholesky.py", line 196, in cho_solve
    b1 = asarray_chkfinite(b)
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\site-packages\numpy\lib\function_base.py", line 498, in asarray_chkfinite
    "array must not contain infs or NaNs")
ValueError: array must not contain infs or NaNs
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "Xx/PycharmProjects/pySIP/examples/thermal_network/armadillo/fit_bayesian.py", line 55, in <module>
    fit = reg.fit(df=df, inputs=inputs, outputs=outputs, options={'init': 'prior', 'hpd': 0.5})
  File "Xx\PycharmProjects\pySIP\pysip\regressors\bayesian.py", line 76, in fit
    uchains, stats, options = dhmc.sample(q0, n_draws, n_chains, n_warmup, options)
  File "Xx\PycharmProjects\pySIP\pysip\mcmc\hmc.py", line 192, in sample
    for n in range(n_chains)
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\site-packages\joblib\parallel.py", line 934, in __call__
    self.retrieve()
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\site-packages\joblib\parallel.py", line 833, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\site-packages\joblib\_parallel_backends.py", line 521, in wrap_future_result
    return future.result(timeout=timeout)
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\concurrent\futures\_base.py", line 428, in result
    return self.__get_result()
  File "Xx\AppData\Local\Continuum\anaconda3\envs\pySIP\lib\concurrent\futures\_base.py", line 384, in __get_result
    raise self._exception
ValueError: array must not contain infs or NaNs
Chain 3:   0%|          | 0/2000 [23:44<?, ?it/s]
Chain 2:   0%|          | 0/2000 [23:44<?, ?it/s]
Chain 1:   0%|          | 0/2000 [23:44<?, ?it/s]
Chain 0:   0%|          | 0/2000 [23:44<?, ?it/s]

**Comment by mjanv** [comment]
Your first error seems to be related to this known issue in tqdm: https://github.com/tqdm/tqdm/issues/613. As a workaround, i will try to install a tqdm version prior to 4.20.

**Comment by Sarah-Ju** [comment]
Thank you @sthebault  for raising this issue I've been having for days and days !

@mjanv  the proposed fix to use tqdm < 4.20 (i.e. 4.19 ?) does not work, hmc.py apparently uses tqdm.auto which does not seem to exist :

  File "/home/sarah/anaconda3/lib/python3.7/site-packages/pysip-0.5-py3.7.egg/pysip/regressors/__init__.py", line 2, in <module>
  File "/home/sarah/anaconda3/lib/python3.7/site-packages/pysip-0.5-py3.7.egg/pysip/regressors/bayesian.py", line 10, in <module>
  File "/home/sarah/anaconda3/lib/python3.7/site-packages/pysip-0.5-py3.7.egg/pysip/mcmc/__init__.py", line 2, in <module>
  File "/home/sarah/anaconda3/lib/python3.7/site-packages/pysip-0.5-py3.7.egg/pysip/mcmc/hmc.py", line 8, in <module>
ModuleNotFoundError: No module named 'tqdm.auto'

I have come to think that this error occurs when one or all the chains terminate(s) prematurely because an inf has been encountered.
Happens to me for example when I use poor priors, always at the initialization of the chains (same as @sthebault ), so what is going on at initialization that throws an inf ?

**Comment by LoicRaillon** [comment]
You are right @Sarah-Ju, this error occurs when a chain terminates prematurely. It could happen for many reasons
- You have posterior probability mass on bounds which cause numerical issues (pySIP doesn't have safety for that or warnings yet)
- Poor identifiability, the HMC diverged into region of the parameter space where the state-space integration is not numerically stable.
- The Fréchet derivative may fail if the state matrix is poorly conditioned and inversely, numerical inaccuracy due to Fréchet derivative may cause divergence in the leapfrog integrator.
- etc.
I have solutions to fix these issues in the future.
